pyoneer/rl/targets/reward_ops.py

```diff
@@ -0,0 +1,668 @@
+            'discounts rank must be <= rewards rank.')
+        discounts = tf.broadcast_to(discounts, rewards_shape)
+
+        # Check lambdas shape, broadcast to rewards shape.
+        lambdas = tf.convert_to_tensor(lambdas)
+        assert lambdas.shape.rank <= rewards.shape.rank, (
+            'lambdas rank must be <= rewards rank.')
+        lambdas = tf.broadcast_to(lambdas, rewards_shape)
+
+        # Check weights shape, broadcast to weights shape.
+        weights = tf.convert_to_tensor(weights)
+        assert weights.shape.rank <= rewards.shape.rank, (
+            'weights rank must be <= rewards rank.')
+        weights = tf.broadcast_to(weights, rewards_shape)
+
+        if bootstrap_value is None:
+            if time_major:
+                bootstrap_value = tf.zeros_like(values[-1, :])
+            else:
+                bootstrap_value = tf.zeros_like(values[:, -1])
+        else:
+            bootstrap_value = tf.convert_to_tensor(bootstrap_value)
+
+        bootstrap_value_t = tf.expand_dims(bootstrap_value, axis=time_axis)
+
+        # Shift values to tp1.
+        if time_major:
+            values_tp1 = values[1:, :]
+        else:
+            values_tp1 = values[:, 1:]
+
+        # Pad the values with the last value.
+        values_tp1 = tf.concat([values_tp1, bootstrap_value_t],
+                               axis=time_axis)
+
+        delta = rewards + discounts * values_tp1 * (1. - lambdas)
+
+        def reduce_fn(agg, cur):
+            next_agg = cur[0] + cur[1] * agg
+            return next_agg
+
+        if time_major:
+            elements = (delta * weights,
+                        lambdas * discounts * weights)
+        else:
+            rank_list = list(range(2, rewards_rank))
+            elements = (tf.transpose(delta * weights, [1, 0] + rank_list),
+                        tf.transpose(lambdas * discounts * weights,
+                                     [1, 0] + rank_list))
+
+        returns = tf.scan(
+            fn=reduce_fn,
+            elems=elements,
+            initializer=tf.zeros_like(bootstrap_value),
+            parallel_iterations=1,
+            back_prop=back_prop,
+            reverse=True)
+
+        if not time_major:
+            rank_list = list(range(2, rewards_rank))
+            returns = tf.transpose(returns, [1, 0] + rank_list)
+        returns = returns * weights
+        returns = tf.debugging.check_numerics(returns, 'Returns')
+        if not back_prop:
+            returns = tf.stop_gradient(returns)
+        return returns
+
+
+@tf.function
+def v_trace_returns(rewards,
+                    values,
+                    log_probs,
+                    log_probs_old,
+                    bootstrap_value=None,
+                    discounts=.99,
+                    weights=1.,
+                    time_major=False,
+                    back_prop=False,
+                    name=None):
+    """Computes v-trace returns.
```

Comment:
What are v-trace returns? Similarly to above. Maybe pointers to papers would be good here, too.